src/test/spec/retryable_reads.rs

```diff
@@ -0,0 +1,183 @@
+        });
+        if has_skipped_op {
+            return;
+        }
+
+        for test_case in test_file.tests {
+            if let Some(skip_reason) = test_case.skip_reason {
+                println!("Skipping {}: {}", test_case.description, skip_reason);
+                continue;
+            }
+
+            let mut options = CLIENT_OPTIONS.clone();
+            if let Some(client_options) = test_case.client_options {
+                options.retry_reads = Some(client_options.get_bool("retryReads").unwrap());
+            }
+            if TestClient::new().await.is_sharded() && test_case.use_multiple_mongoses != Some(true)
+            {
+                options.hosts = options.hosts.iter().cloned().take(1).collect();
+            }
+            let client = EventClient::with_options(options).await;
+
+            if let Some(ref run_on) = test_file.run_on {
+                let can_run_on = run_on.iter().any(|run_on| {
+                    if let Some(ref min_server_version) = run_on.min_server_version {
+                        let (major, minor) = parse_version(min_server_version);
+                        if client.server_version_lt(major, minor) {
+                            return false;
+                        }
+                    }
+                    if let Some(ref max_server_version) = run_on.max_server_version {
+                        let (major, minor) = parse_version(max_server_version);
+                        if client.server_version_gt(major, minor) {
+                            return false;
+                        }
+                    }
+                    if let Some(ref topology) = run_on.topology {
+                        if !topology.contains(&client.topology()) {
+                            return false;
+                        }
+                    }
+                    true
+                });
+                if !can_run_on {
+                    println!("Skipping {}", test_case.description);
+                    continue;
+                }
+            }
+
+            match test_case.fail_point {
+                Some(_) => {
+                    let _guard = LOCK.run_exclusively().await;
+                }
+                _ => {
+                    let _guard = LOCK.run_concurrently().await;
+                }
+            }
+
+            let db_name = match test_file.database_name {
+                Some(ref db_name) => db_name.clone(),
+                None => test_case.description.replace('$', "%").replace(' ', "_"),
+            };
+
+            let coll_name = match test_file.collection_name {
+                Some(ref coll_name) => coll_name.clone(),
+                None => String::from("coll"),
+            };
+
+            let coll = client.init_db_and_coll(&db_name, &coll_name).await;
+
+            if let Some(ref data) = test_file.data {
+                let data: Vec<Document> = data
+                    .as_array()
```

Comment:
This is what I had originally, but ran into a problem with the GridFS tests that required a more generalized type. Even though the GridFS tests don't actually run, they need to be serialized into `TestFile`s in order to check if they contain an operation that should be skipped. GridFS tests represent their data as a [document](https://github.com/mongodb/specifications/blob/8b1d6e9ff4a7dfd2a47d3fcf6e0ee6b2d98291a7/source/retryable-reads/tests/gridfs-download-serverErrors.json#L19) whereas the rest of the tests have data as an [array of documents](https://github.com/mongodb/specifications/blob/8b1d6e9ff4a7dfd2a47d3fcf6e0ee6b2d98291a7/source/retryable-reads/tests/findOne.json#L19), so the type that represents data needs to be general enough to support both formats. I considered using an enum to represent the two types, but that complicated serialization so I abstracted it out to `Bson`.

Reply:
Ah good point. Then we can use an enum to model this and instruct serde how to deserialize it via the `untagged` attribute (https://serde.rs/enum-representations.html#untagged).

something like:
```rust
#[derive(Deserialize)]
#[serde(untagged)]
enum TestData {
    Single(Vec<Document>),
    Many(HashMap<String, Vec<Document>>)
}
```
Because Rust has such a strict type system, working with JSON/BSON can get _very_ boilerplate heavy unless you model the data precisely. The nice thing is that the more specific your models are, the more boilerplate goes away and is handled by `serde`.

Reply:
> Because Rust has such a strict type system, working with JSON/BSON can get very boilerplate heavy unless you model the data precisely. The nice thing is that the more specific your models are, the more boilerplate goes away and is handled by serde

I'm totally going to steal this quote at some point